cf-revision-manager.php

```diff
@@ -2,200 +2,321 @@
+}
+
+/**
+ * Register js and css file.
+ *
+ * @since 1.0.1
+ */
+function cfrm_enqueue_js() {
+	wp_register_script( 'cfrm-script', cfrm_get_global( 'directory_url' ) . 'cfrm-script.js', array( 'jquery' ) );
+	wp_register_style( 'cfrm-style', cfrm_get_global( 'directory_url' ) . 'cfrm-style.css' );
+}
+add_action( 'admin_enqueue_scripts', 'cfrm_enqueue_js' );
+
+/**
+ * Enqueue js and css file in revision page.
+ *
+ * @since  1.0.1
+ * @return void
+ */
+function cfrm_admin_head() {
+	if ( ! empty( $_GET['page'] ) && $_GET['page'] == basename( __FILE__ ) ) {
+		wp_enqueue_style( 'cfrm-style' );
+		wp_enqueue_script( 'cfrm-script' );
 	}
+}
+add_action( 'admin_head', 'cfrm_admin_head' );
+
+/**
+ * Register revision manger admin menu.
+ *
+ * @since  1.0.1
+ * @return void
+ */
+function cfrm_admin_menu() {
+	 add_options_page(
+        __( 'Revision Manager', 'cfrm' ),
+        __( 'Revision Manager', 'cfrm' ),
+        'manage_options',
+        basename( __FILE__ ),
+        'cfrm_admin_form'
+    );
+}
+add_action( 'admin_menu', 'cfrm_admin_menu' );
 
-	class cf_revisions {
-		private static $_instance;
-		protected $postmeta_keys = array();
-	
-		public function __construct() {
-			# save & restore
-			add_action('save_post', array($this, 'save_post_revision'), 10, 2);
-			add_action('wp_restore_post_revision', array($this, 'restore_post_revision'), 10, 2);
-
-			if (is_admin()) {		
-				# revision display
-				global $pagenow;
-				if ($pagenow == 'revision.php') {
-					add_filter('_wp_post_revision_fields', array($this, 'post_revision_fields'), 10, 1);
-					add_filter('_wp_post_revision_field_postmeta', array($this, 'post_revision_field'), 1, 2);
+/**
+ * Revision Manger admin form callback function.
+ *
+ * @since  1.0.1
+ * @return void
+ */
+function cfrm_admin_form() {
+	//include_once cfase_get_global( 'directory_path' ) . 'includes/admin-form.php';
+	global $CFR_KEYS_REQUIRED;
+	$required_keys = $CFR_KEYS_REQUIRED;
+	$keys = array_diff( cfrm_meta_keys(), excluded_keys(), $required_keys );
+?>
+<div class="wrap">
+	<h2><?php _e( 'CF Revision Manager', 'cfrm' ); ?></h2>
+<?php
+		if ( ! count( $keys ) ) {
+			echo '<p>'.__( 'No custom fields found.', 'cfrm' ).'</p>';
+		}
+		else {
+			echo '<form id="cfr_revision_manager_form" name="cfr_revision_manager_form" action="'.admin_url('options-general.php').'" method="post">
+				<p>'.__( 'A plugin or theme has specified that the following custom fields need to included in revisions.', 'cfrm' ).'</p>';
+			if ( count( $required_keys ) ) {
+				echo '<div>
+				<ul id="cfr_revision_manager_keys_required">';
+				foreach ($required_keys as $key) {
+					$checked = $key;
+					$disabled = $key;
+					$id = 'cf_revision_manager_key_'.esc_attr($key);
+					echo '<li>
+						<input type="checkbox" name="revision_manager_keys[]" id="'.$id.'" value="'.esc_attr($key).'" '.checked($key, $checked, false).' '.disabled($key, $disabled, false).' />
```

Comment:
Use `sprintf()` instead of `echo`. The `$id` variable should be escaped for output (`esc_attr()`). This is technically safe written as-is because we do escape the one part of the id that is dynamic, but it reads better (and is a better practice) to escape the entire variable late rather than to escape part of it early..